### DISCUSSION TITLE
Terminal.Gui doesn't support AOT

### DISCUSSION THREAD

**Discussion by chenjing1294** [tui-cs/Terminal.Gui#2414] [General]
Are there plans to support AOT?

**Comment by BDisp** [comment]
I think you are meaning about `AOT (Ahead-of-time compilation)`. I don't have enough knowledge about that, but I think it only supports `JIT (Just-in-time compilation)`.

**Comment by tznind** [comment]
If you add the following to UICatalog.csproj and compile with `dotnet publish -r win-x64 -c Release` then you get 31 warnings:

![image](https://user-images.githubusercontent.com/31306100/225372583-555d11ae-fb6a-49d1-bdec-f72db0e989ef.png)

But the resulting binary does run and doesn't seem to crash at all when I navigate around it.  If these warnings can be addressed that would be great.  Any way I could tempt you into submitting a PR to address these @chenjing1294 ?

Alternatively if you could go into more detail on why you need AoT or any specific (non warning) issues you've had using AOT with Terminal.Gui we could perhaps address those?

```csproj
  <PropertyGroup>
    <PublishAot>true</PublishAot>	
    <StripSymbols>true</StripSymbols>
</PropertyGroup>
```

**Reply by chenjing1294** [reply]
But running on my computer will report an error:(develop or v2_develop branch)
```
PS C:\Users\chenjing\GitRepository\Terminal.Gui\UICatalog\bin\x64\Release\net7.0\win-x64\publish> .\UICatalog.exe
Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at UICatalog.UICatalogApp.Main(String[]) + 0x122
   at UICatalog!<BaseAddress>+0x2bf1eb
```

**Reply by tznind** [reply]
Ah I think I was running the wrong binary.  I see your issue.   What is your goal? AOT is usually used for webapps and is about fast startup and performance critical situations (e.g. microservices/kubernetes etc).  

If the goal is to ship a single executable that does not require a framework install then you can instead use publish single file.

```
<PropertyGroup>
	<PublishSingleFile>true</PublishSingleFile>
	<RuntimeIdentifiers>win-x64</RuntimeIdentifiers>
</PropertyGroup>
```

I have created a PR (https://github.com/gui-cs/Terminal.Gui/pull/2422) which fixes the one error I can see when runing under those settings.  The result is a 68MB binary.

There are 3 actions to supporting AOT if thats still a requirement.  [AOT forces trimming](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/#limitations-of-native-aot-deployment) and imo that is a barrier for UICatalog since it leans heavily on reflection to build/discover Scenarios.

- [x] Support `PublishSingleFile` in UICatalog (see [my PR](https://github.com/gui-cs/Terminal.Gui/pull/2422))
- [ ] Supporting `PublishTrimmed` / `AOT` in gui.cs itself (e.g. proof of concept example app) 
- [ ] Supporting `PublishTrimmed` in UICatalog (requires rewritting all the reflection)
- [ ] Supporting AOT (probably this will just work after the above 2 are done)

Here is a small investigation into what supporting `PublishTrimmed` would involve.  WIP-https://github.com/gui-cs/Terminal.Gui/pull/2423
